### PR TITLE
ISSUE-1040 Implement CassandraPublicAssetRepository

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve https://github.com/linagora/tmail-backend/issues/1040

### CODE REVIEW

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/CassandraPublicAssetRepositoryDAO.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  val BLOB_ID: CqlIdentifier = CqlIdentifier.fromCql("blob_id")
+  val SIZE: CqlIdentifier = CqlIdentifier.fromCql("size")
+  val CONTENT_TYPE: CqlIdentifier = CqlIdentifier.fromCql("content_type")
+  val IDENTITY_IDS: CqlIdentifier = CqlIdentifier.fromCql("identity_ids")
+
+  val MODULE: CassandraModule = CassandraModule.table(TABLE_NAME)
+    .comment("Hold user public assets metadata")
+    .statement(statement => _ => statement
+      .withPartitionKey(USER, TEXT)
+      .withClusteringColumn(ASSET_ID, DataTypes.UUID)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/CassandraPublicAssetRepositoryDAO.scala:38]: TimeUUID?

**Review by Arsnael — APPROVED** [review]
Read it nothing to add

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAsset.scala [hunk 2]
```diff
@@ -15,7 +16,7 @@ import org.apache.james.mailbox.model.ContentType
 import scala.util.Try
 
 object PublicAssetIdFactory {
-  def generate(): PublicAssetId = PublicAssetId(UUID.randomUUID())
+  def generate(): PublicAssetId = PublicAssetId(Generators.timeBasedGenerator().generate())
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAsset.scala:19]: Do we really need a new lib for that? How were we doing until now for time uuids?

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAsset.scala:19]: We use Cassandra driver util class to generate time uuid. But this is the API layer so we can not leverage Cassandra driver one.

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAsset.scala:19]: At that cost I prefer using UUID than time UUID for this.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/CassandraPublicAssetRepositoryDAO.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      .withColumn(SIZE, BIGINT)
+      .withColumn(CONTENT_TYPE, TEXT)
+      .withColumn(IDENTITY_IDS, frozenSetOf(DataTypes.UUID)))
+    .build
+
+  val FROZEN_OF_UUIDS_CODEC: TypeCodec[java.util.Set[UUID]] = CodecRegistry.DEFAULT.codecFor(frozenSetOf(DataTypes.UUID))
+  val LIST_OF_UUIDS_CODEC: TypeCodec[java.util.List[UUID]] = CodecRegistry.DEFAULT.codecFor(listOf(DataTypes.UUID))
+}
+
+class CassandraPublicAssetRepositoryDAO @Inject()(session: CqlSession) {
```
Comment (hungphan227) [inline tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/CassandraPublicAssetRepositoryDAO.scala:50]: should be CassandraPublicAssetDAO

**Comment by quantranhong1999** [comment]
I have reworked a bit following the latest API change.

We can leverage easily `TIMEUUID` for `ASSET_ID` now.

**Review by Arsnael — APPROVED** [review]

**Review by hungphan227 — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Pull request merged**